File: LeetCode/Easy/1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        return dfs(root, 0);
    }
    
    int dfs(TreeNode* node, int curr){
        if (!node->left && !node->right) return curr * 2 + node->val;
        
        int ret = 0;
        curr = curr * 2 + node->val;
        
        if (node->left) ret += dfs(node->left, curr);
        if (node->right) ret += dfs(node->right, curr);
        
        return ret;
    }
};

File: LeetCode/Easy/1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumRootToLeaf(root *TreeNode) int {
    dfs := func(node *TreeNode, curr int) int { return 0 }
    dfs = func(node *TreeNode, curr int) int {
        if node.Left == nil && node.Right == nil { return curr * 2 + node.Val }
        
        ret := 0
        curr = curr * 2 + node.Val
        
        if node.Left != nil { ret += dfs(node.Left, curr) }
        if node.Right != nil { ret += dfs(node.Right, curr) }
        
        return ret
    }
    
    return dfs(root, 0)
}